### PR TITLE
Visualize chained controllers with graphviz

### PR DESCRIPTION
**Pull request by pac48**
This change depends on this [PR](https://github.com/ros-controls/ros2_control/pull/758). It adds a new command line option `ros2 control view_controller_chains`, which generates an image using Graphviz.   
Screenshot from 2022-07-02 23-56-55

### CODE REVIEW

**Comment by destogl** [comment]
This looks very cool! What do you think that we create a rqt plugin for that?

**Comment by pac48** [comment]
> This looks very cool! What do you think that we create a rqt plugin for that?

I was taking inspiration from `ros2 run tf2_tools view_frames.py` described [here](https://docs.ros.org/en/foxy/Tutorials/Intermediate/Tf2/Introduction-To-Tf2.html), but I think an rqt_plugin would also be a good idea.

**Comment by ejalaa12** [comment]
That is a very cool feature indeed.
In general a plugin that displays the state of each controller(active, inactive, hardware interface claimed) with features like the one that existed in rqt in Ros1 would be very useful.

**Comment by pac48** [comment]
> That is a very cool feature indeed. In general a plugin that displays the state of each controller(active, inactive, hardware interface claimed) with features like the one that existed in rqt in Ros1 would be very useful.

What if I make the lines dashed if the interface is not claimed and solid if it is? And what if I put the controller status (active, inactive) underneath the name at the top of each node?

**Comment by ejalaa12** [comment]
> What if I make the lines dashed if the interface is not claimed and solid if it is? And what if I put the controller status (active, inactive) underneath the name at the top of each node?

Agreed for the dashed lines.
I would add a either a small circle (red for inactive, green for active) on the side, or just change the text color, this would be more readable than having an additional text to tell if it is active inactive.

**Review by destogl — APPROVED** [review]

**Review by bmagyar — APPROVED** [review]

File: ros2controlcli/package.xml [hunk 1]
```diff
@@ -18,6 +18,7 @@
   <depend>controller_manager</depend>
   <depend>controller_manager_msgs</depend>
   <exec_depend>rosidl_runtime_py</exec_depend>
+  <exec_depend>graphviz</exec_depend>
```
Comment (bmagyar) [inline ros2controlcli/package.xml:21]:
```suggestion
  <exec_depend>python3-pygraphviz</exec_depend>
```

File: ros2controlcli/package.xml [hunk 2]
```diff
@@ -18,6 +18,7 @@
   <depend>controller_manager</depend>
   <depend>controller_manager_msgs</depend>
   <exec_depend>rosidl_runtime_py</exec_depend>
+  <exec_depend>python3-pygraphviz</exec_depend>
```
Comment (bmagyar) [inline ros2controlcli/package.xml:21]:
```suggestion
  <exec_depend>python-graphviz-pip</exec_depend>
```

File: ros2controlcli/package.xml [hunk 3]
```diff
@@ -18,6 +18,7 @@
   <depend>controller_manager</depend>
   <depend>controller_manager_msgs</depend>
   <exec_depend>rosidl_runtime_py</exec_depend>
+  <exec_depend>python-graphviz-pip</exec_depend>
```
Comment (bmagyar) [inline ros2controlcli/package.xml:21]:
```suggestion
  <exec_depend>python-graphviz-pip</exec_depend>
  <exec_depend>graphviz</exec_depend>
```

File: ros2controlcli/ros2controlcli/verb/view_controller_chains.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+from ros2cli.node.strategy import NodeStrategy
+from ros2cli.verb import VerbExtension
+
+from ros2controlcli.api import add_controller_mgr_parsers
+
+import graphviz
+
+
+# controller_name, state_connections[controller_name], command_connections[controller_name],
+# input_chain_connections[controller_name], output_chain_connections
```
Comment (bmagyar) [inline ros2controlcli/ros2controlcli/verb/view_controller_chains.py:28]:
```suggestion
```

File: ros2controlcli/ros2controlcli/verb/view_controller_chains.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            s.edge('{}:{}'.format(controller_name, "controller_start_" + connection),
+                   '{}:{}'.format(port_map['controller_end_' + connection], 'controller_end_' + connection))
+        for state_connection in state_connections[controller_name]:
+            s.edge('{}:{}'.format("state_interfaces", "state_start_" + state_connection),
+                   '{}:{}'.format(controller_name, 'state_end_' + state_connection))
+        for command_connection in command_connections[controller_name]:
+            s.edge('{}:{}'.format(controller_name, "command_start_" + command_connection),
+                   '{}:{}'.format("command_interfaces", 'command_end_' + command_connection))
+
+    # s.attr(splines="false")
```
Comment (bmagyar) [inline ros2controlcli/ros2controlcli/verb/view_controller_chains.py:122]:
```suggestion
```

File: ros2controlcli/test/test_view_controller_chains.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+            chained_to_controller.claimed_interfaces.append(f"joint{i}/position")
+            chained_to_controller.claimed_interfaces.append(f"joint{i}/velocity")
+            chained_to_controller.required_state_interfaces.append(f"joint{i}/position")
+            chained_to_controller.reference_interfaces.append(f"joint{i}/position")
+            chained_to_controller.reference_interfaces.append(f"joint{i}/velocity")
+
+            # chained_from_controller.claimed_interfaces.append(f"{chained_to_controller.name}/joint{i}/position")
+            # chained_from_controller.claimed_interfaces.append(f"{chained_to_controller.name}/joint{i}/velocity")
+            chained_from_controller.required_state_interfaces.append(f"joint{i}/position")
```
Comment (bmagyar) [inline ros2controlcli/test/test_view_controller_chains.py:61]:
```suggestion
            chained_from_controller.required_state_interfaces.append(f"joint{i}/position")
```

**Review by bmagyar — CHANGES_REQUESTED** [review]
Minor cleanup

**Review by bmagyar — APPROVED** [review]

**Pull request merged**